src/Ledger/TokenAlgebra/ValueRelation.lagda

```diff
@@ -0,0 +1,464 @@
+
+⊕₁-assoc : Algebra.Associative _⇐⇒_ _⊕₁_
+⊕₁-assoc Ru Rv Rw = i , ii
+ where
+ i : ((Ru ⊕₁ Rv) ⊕₁ Rw) ⇒ (Ru ⊕₁ (Rv ⊕₁ Rw))
+ i {aid} {q} (quv , qw , (qu , qv , Ruqu , Rvqv , quv≡quqv) , Rwqw , q≡quvqw) = Goal
+  where
+  assc : q ≡ qu + (qv + qw)
+  assc = begin  q               ≡⟨ q≡quvqw ⟩
+                quv + qw        ≡⟨ cong (λ x → x + qw) quv≡quqv ⟩
+                qu + qv + qw    ≡⟨ +-assoc qu qv qw ⟩
+                qu + (qv + qw)  ∎
+
+  Goal : (Ru ⊕₁ (Rv ⊕₁ Rw)) aid q
+  Goal = qu , qv + qw , Ruqu , (qv , qw , Rvqv , Rwqw , refl) , assc
+
+ ii : (Ru ⊕₁ (Rv ⊕₁ Rw)) ⇒ ((Ru ⊕₁ Rv) ⊕₁ Rw)
+ ii {aid} {q} (qu , qvw , Ruqu , (qv , qw , Rvqv , Rwqw , qvw≡qvqw) , q≡quqvw) = Goal
+  where
+  assc : qu + qvw ≡ qu + qv + qw
+  assc = begin  qu + qvw        ≡⟨ cong (λ x → qu + x) qvw≡qvqw ⟩
+                qu + (qv + qw)  ≡⟨ sym (+-assoc qu qv qw) ⟩
+                qu + qv + qw    ∎
+
+  Goal : ((Ru ⊕₁ Rv) ⊕₁ Rw) aid q
+  Goal = qu + qv , qw , (qu , qv , Ruqu , Rvqv , refl) , Rwqw , trans q≡quqvw assc
+
+
+⊕-assoc : Algebra.Associative _≋_ _⊕_
+⊕-assoc (Ru , luu) (Rv , lvu) (Rw , lwu)  = i , ii
+ where
+ i : (Ru ⊕₁ Rv) ⊕₁ Rw ⇒ Ru ⊕₁ (Rv ⊕₁ Rw)
+ i {aid} {q} (quv , qw , (qu , qv , Ruqu , Rvqv , quv≡quqv) , Rwqw , q≡quvqw) = Goal
+  where
+  assc : quv + qw ≡ qu + (qv + qw)
+  assc = begin  quv + qw        ≡⟨ cong (λ x → x + qw) quv≡quqv ⟩
+                qu + qv + qw    ≡⟨ +-assoc qu qv qw ⟩
+                qu + (qv + qw)  ∎
+
+  Goal : (Ru ⊕₁ (Rv ⊕₁ Rw)) aid q
+  Goal = qu , ((qv + qw) , (Ruqu , ((qv , qw , Rvqv , Rwqw , refl) , trans q≡quvqw assc )))
+
+
+ ii : Ru ⊕₁ (Rv ⊕₁ Rw) ⇒ (Ru ⊕₁ Rv) ⊕₁ Rw
+ ii {aid} {q} (qu , qvw , Ruqu , (qv , qw , Rvqv , Rwqw , qvw≡qvqw) , q≡quqvw) = Goal
+  where
+  tras : q ≡ qu + qv + qw
+  tras = begin  q               ≡⟨ q≡quqvw ⟩
+                qu + qvw        ≡⟨ cong (λ x → qu + x) qvw≡qvqw ⟩
+                qu + (qv + qw)  ≡⟨ sym (+-assoc qu qv qw) ⟩
+                qu + qv + qw    ∎
+
+  Goal : ((Ru ⊕₁ Rv) ⊕₁ Rw) aid q
+  Goal = (qu + qv) , (qw , ((qu , qv , Ruqu , Rvqv , refl) , (Rwqw , tras)))
+
+
+⊕-cong : Algebra.Congruent₂ _≋_ _⊕_
+⊕-cong {Ru , luu} {Rv , lvu} {Ru' , luu'} {Rv' , lvu'} (Ru⇒Rv , Rv⇒Ru) (Ru'⇒Rv' , Rv'⇒Ru') = i , ii
+ where
+ i : (Ru ⊕₁ Ru') ⇒ (Rv ⊕₁ Rv')
+ i {aid} {q} (qu , qu' , Ruqu , Ru'qu' , q≡ququ') = qu , qu' , Ru⇒Rv Ruqu , Ru'⇒Rv' Ru'qu' , q≡ququ'
+
+ ii : (Rv ⊕₁ Rv') ⇒ (Ru ⊕₁ Ru')
+ ii {aid} {q} (qv , qv' , Rvqv , Rv'qv' , q≡qvqv') = qv , qv' , Rv⇒Ru Rvqv , Rv'⇒Ru' Rv'qv' , q≡qvqv'
+\end{code}
+
+\subsection{Definition of the value monoid}
+
+An inhabitant of `Value` is a map denoting a finite collection of quantities of assets.
+
+\begin{code}
+open TokenAlgebraPoly
+open CommutativeMonoid renaming (_∙_ to _⋆_) hiding (trans ; sym)
+open Algebra
+
+module _
+  {AdaPolicy   : PolicyID}
+  {DecEqValue  : DecidableEquality (AssetID ⇀ Quantity) }
+  {Size        : AssetID ⇀ Quantity → ℕ}
+  {AdaForAll   : (R : AssetID ⇀ Quantity) → ∃[ q ] (proj₁ R) (AdaPolicy , AdaName) q}
```

Comment:
Note that `AdaForAll` implies `⊥`:
```
contr : ⊥
contr = proj₂ $ AdaForAll ((λ _ _ → ⊥) , λ where () ())
```
To not have a contradiction, you'd need to instead define a subset of relations that satisfy this property. However, everything in this file becomes way easier if you instead require all relations to be total. Then, the definition of `⊕₁` wouldn't require the two existentials for example. This also fixes the issue that there are tons of projections all over the place.